azure/ConsiderationsForServiceDesign.md

```diff
@@ -229,6 +229,33 @@ An important distinction between RELO and status monitor LROs is that there is a
 status monitor LRO, whereas the status of all RELO operations is combined into the status of the resource.
 So status monitor LROs are "one-to-one" with their operation status, whereas RELO-style LROs are "many-to-one".
 
+## Errors
+One of the most important parts of service design is also one of the most overlooked.  The errors returned by your service are a critical part of your developer experience.  Your service and your customer's application together form a distributed system.  Errors are inevitable, but well-designed errors can help you avoid costly customer support incidents by empowering customers to self-diagnose problems.
```

Comment:
Should we consider adding a link to more documentation so customers can self-diagnose?  If we include a link, should we use Markdown syntax?  Here is an example from Azure ML.

![image](https://user-images.githubusercontent.com/1988517/146137313-e33a489a-807c-46c9-88e5-c711887b220e.png)

Reply:
I've got a lot of thoughts about troubleshooting documentation that are outside the scope of these changes.  If we have additional pointers today, I personally don't like the idea of using markdown because I want error messages to be as human readable as possible.  Seeing `You can find troubleshooting steps [here](https://www.azurehelp.net/storage/blobs/blob_delete/404/BlobNotFound)` doesn't do more than `You can find troubleshooting steps at https://www.azurehelp.net/storage/blobs/blob_delete/404/BlobNotFound`.  I know at least VS and VS Code make hyperlinks in error messages easily clickable as well.

Reply:
To confirm, we can/should include hyperlinks to external content in error messages.  But we should use plain URL and let the UI/tool render them as clickable links.  One argument for using plain URL is that UI/tool would need to parse Markdown otherwise.  And only parsing URLs would be awkward.

Reply:
@bojunehsu OpenAPI explicitly allows markdown in many of the `description` elements of the doc -- I don't think we should advise against using it. Rather, if our tooling is lacking in support of markdown then lets identify those gaps and see if we can get them fixed.

Reply:
We are currently referring to the custom-defined _message_ property, which doesn't need to follow OpenAPI conventions.

On the separate topic of markdown in Swagger descriptions, below are places where our tooling currently does not support markdown:
* API Management - We can import APIs definitions from Markdown.  If I recall correctly, the imported documentation does not properly render markdown.
* Autorest - The description properties are used to populate the documentation rendered in tooltips.  VS tooltips does not support Markdown (it would be very awkward).  I don't recall how markdown documentation is rendered in the source code comments.
* JSON Editor - Viewing/editing markdown in JSON editors is not a pleasant experience.  The entire markdown needs to be in a single line, using escaped newlines.   It is nearly usable.  YAML has better Markdown support.

Reply:
https://github.com/Azure/autorest.csharp/issues/938 is tracking translation of markdown to C# doc comments in AutoRest - though we need similar support across languages.

Reply:
The updated guidance is good. I don't think we should return formatting, MD, HTML, etc. in the returned text.